File: leetcode/Linked List Cycle.py
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, x):
#         self.val = x
#         self.next = None

class Solution:
    # @param head, a ListNode
    # @return a boolean
    def hasCycle(self, head):
        if not head:
            return False
        d = {}
        p = head
        while p:
            if d.has_key(p):
                return True
            else:
                d[p] = 1
            p = p.next
        return False

File: lintcode/Linked_List_Cycle.cpp
/**
 * Definition of ListNode
 * class ListNode {
 * public:
 *     int val;
 *     ListNode *next;
 *     ListNode(int val) {
 *         this->val = val;
 *         this->next = NULL;
 *     }
 * }
 */
class Solution {
public:
    /**
     * @param head: The first node of linked list.
     * @return: True if it has a cycle, or false
     */
    bool hasCycle(ListNode *head) {
        if (NULL == head)
            return false;
        
        ListNode dummy(0);
        dummy.next = head;
        head = &dummy;
        ListNode *fast = head;
        ListNode *slow = head;
        while (fast && slow) {
            if (NULL == fast->next) {
                return false;
            }
            fast = fast->next->next;
            slow = slow->next;
            if (fast == slow)
                return true;
        }
        return false;
    }
};



